server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -166,33 +166,54 @@ public CachingClusteredClient(
+    // and might blow up in some cases https://github.com/druid-io/druid/issues/2108
+    int populateUncoveredIntervals = BaseQuery.getContextPopulateUncoveredIntervals(query, 0);
+
+    if (populateUncoveredIntervals > 0) {
+      List<Interval> uncoveredIntervals = Lists.newArrayListWithCapacity(populateUncoveredIntervals);
+      boolean uncoveredIntervalsOverflowed = false;
+
+      for (Interval interval : query.getIntervals()) {
+        Iterable<TimelineObjectHolder<String, ServerSelector>> lookup = timeline.lookup(interval);
+        long startMillis = interval.getStartMillis();
+        long endMillis = interval.getEndMillis();
+        for (TimelineObjectHolder<String, ServerSelector> holder : lookup) {
+          Interval holderInterval = holder.getInterval();
+          long intervalStart = holderInterval.getStartMillis();
+          if (!uncoveredIntervalsOverflowed && startMillis != intervalStart) {
+            if (populateUncoveredIntervals > uncoveredIntervals.size()) {
+              uncoveredIntervals.add(new Interval(startMillis, intervalStart));
+            } else {
+              uncoveredIntervalsOverflowed = true;
+            }
+          }
+          startMillis = holderInterval.getEndMillis();
+          serversLookup.add(holder);
         }
-        startMillis = holderInterval.getEndMillis();
-        serversLookup.add(holder);
-      }
 
-      if (startMillis < endMillis) {
-        uncoveredIntervals.add(new Interval(startMillis, endMillis));
+        if (!uncoveredIntervalsOverflowed && startMillis < endMillis) {
+          if (populateUncoveredIntervals > uncoveredIntervals.size()) {
+            uncoveredIntervals.add(new Interval(startMillis, endMillis));
+          } else {
+            uncoveredIntervalsOverflowed = true;
+          }
+        }
       }
-    }
 
-    if (!uncoveredIntervals.isEmpty()) {
-      // This returns intervals for which NO segment is present.
-      // Which is not necessarily an indication that the data doesn't exist or is
-      // incomplete. The data could exist and just not be loaded yet.  In either
-      // case, though, this query will not include any data from the identified intervals.
-      responseContext.put("uncoveredIntervals", uncoveredIntervals);
+      if (!uncoveredIntervals.isEmpty()) {
+        // This returns intervals for which NO segment is present.
+        // Which is not necessarily an indication that the data doesn't exist or is
+        // incomplete. The data could exist and just not be loaded yet.  In either
+        // case, though, this query will not include any data from the identified intervals.
+        responseContext.put("uncoveredIntervals", uncoveredIntervals);
+        responseContext.put("uncoveredIntervalsOverflowed", uncoveredIntervalsOverflowed);
+      }
+    } else {
+      for (Interval interval : query.getIntervals()) {
+        Iterables.addAll(serversLookup, timeline.lookup(interval));
```

Comment:
there is only one call ... the code flows are different when uncoveredIntervals are to be populated and not populated. its simpler this way bcoz the code for uncoveredIntervals disabled (else on line 213)  is simple and better on its own.